src/box/memtx_tx.c

```diff
@@ -1318,6 +1318,24 @@ memtx_tx_history_rollback_stmt(struct txn_stmt *stmt)
 			assert(newer->link[i].older_story == older);
 			assert(older == NULL ||
 			       older->link[i].newer_story == newer);
+
+			if (i != 0 || older == NULL)
+				continue;
+
+			/**
+			 * Relink those who delete story and
+			 * make them delete older story.
+			 */
+			while (story->del_stmt != NULL) {
+				struct txn_stmt *next =
+					story->del_stmt->next_in_del_list;
+				story->del_stmt->next_in_del_list =
+					older->del_stmt;
+				older->del_stmt = story->del_stmt;
+				story->del_stmt->del_story = older;
+				older->del_stmt = newer->add_stmt;
+				story->del_stmt = next;
```

Comment:
It's quite misleading that you change story->del_stmt and uses it in while cycle. Why not declare separate variable and use it as iterator? In the end simply nullify story->del_stmt

Reply:
Finally, AFAIU this can be done in a separate patch...